File: code/c++/958. 二叉树的完全性检验.cpp
/**
 * 958. Check Completeness of a Binary Tree
 * Definition for a binary tree node.
 * struct TreeNode {
 *     int val;
 *     TreeNode *left;
 *     TreeNode *right;
 *     TreeNode(int x) : val(x), left(NULL), right(NULL) {}
 * };
 * time:O(N)
 */
class Solution {
public:
  bool isCompleteTree(TreeNode *root) {
    if (NULL == root)
      return true;
    int nodes = nubmerOfNodes(root);
    return dfs(root, 1, nodes);
  }

  bool dfs(TreeNode *root, int level, int &total) {
    if (NULL == root)
      return true;
    if (level > total)
      return false;
    //完全二叉树的性质
    bool isComp = dfs(root->left, 2 * level, total) &&
                  dfs(root->right, 2 * level + 1, total);
    return isComp;
  }
  //
  int nubmerOfNodes(TreeNode *root) {
    if (NULL == root)
      return 0;
    return 1 + nubmerOfNodes(root->left) + nubmerOfNodes(root->right);
  }
};

File: code/go/958. 二叉树的完全性检验.go
package main

/**958. Check Completeness of a Binary Tree
 * Definition for a binary tree node.
 * type TreeNode struct {
 *     Val int
 *     Left *TreeNode
 *     Right *TreeNode
 * }
 */
func isCompleteTree(root *TreeNode) bool {
	if root == nil {
		return true
	}
	total := getNodes(root)

	return isCompleteTreeByIndex(root, 1, total)

}

func getNodes(root *TreeNode) int {

	if root == nil {
		return 0
	}

	return 1 + getNodes(root.Left) + getNodes(root.Right)
}

func isCompleteTreeByIndex(root *TreeNode, index int, total int) bool {
	if nil == root {
		return true
	}

	if index > total {
		return false
	}

	return isCompleteTreeByIndex(root.Left, 2*index, total) && isCompleteTreeByIndex(root.Right, 2*index+1, total)
}
